limitador/src/storage/mod.rs

```diff
@@ -21,38 +23,230 @@ pub enum Authorization<'c> {
+    limits: RwLock<HashMap<Namespace, HashSet<Limit>>>,
+    counters: Box<dyn AsyncCounterStorage>,
+}
+
+impl Storage {
+    pub fn new() -> Self {
+        Self {
+            limits: RwLock::new(HashMap::new()),
+            counters: Box::new(InMemoryStorage::default()),
+        }
+    }
+
+    pub fn with_counter_storage(counters: Box<dyn CounterStorage>) -> Self {
+        Self {
+            limits: RwLock::new(HashMap::new()),
+            counters,
+        }
+    }
+
+    pub fn get_namespaces(&self) -> HashSet<Namespace> {
+        self.limits.read().unwrap().keys().cloned().collect()
+    }
+
+    pub fn add_limit(&self, limit: Limit) {
+        let namespace = limit.namespace().clone();
+        self.limits
+            .write()
+            .unwrap()
+            .entry(namespace)
+            .or_default()
+            .insert(limit);
+    }
+
+    pub fn get_limits(&self, namespace: &Namespace) -> HashSet<Limit> {
+        match self.limits.read().unwrap().get(namespace) {
+            Some(limits) => limits.iter().cloned().collect(),
+            None => HashSet::new(),
+        }
+    }
+
+    pub fn delete_limit(&self, limit: &Limit) -> Result<(), StorageErr> {
+        let mut limits = HashSet::new();
+        limits.insert(limit.clone());
+        self.counters.delete_counters(limits)?;
+
+        let mut limits = self.limits.write().unwrap();
+
+        if let Some(counters_by_limit) = limits.get_mut(limit.namespace()) {
+            counters_by_limit.remove(limit);
+
+            if counters_by_limit.is_empty() {
+                limits.remove(limit.namespace());
+            }
+        }
+        Ok(())
+    }
+
+    pub fn delete_limits(&self, namespace: &Namespace) -> Result<(), StorageErr> {
+        if let Some(data) = self.limits.write().unwrap().remove(namespace) {
+            let limits = data.iter().cloned().collect();
+            self.counters.delete_counters(limits)?;
+        }
+        Ok(())
+    }
+
+    pub fn is_within_limits(&self, counter: &Counter, delta: i64) -> Result<bool, StorageErr> {
+        self.counters.is_within_limits(counter, delta)
+    }
+
+    pub fn update_counter(&self, counter: &Counter, delta: i64) -> Result<(), StorageErr> {
+        self.counters.update_counter(counter, delta)
+    }
+
+    pub fn check_and_update<'c>(
+        &self,
+        counters: &HashSet<&'c Counter>,
+        delta: i64,
+    ) -> Result<Authorization<'c>, StorageErr> {
+        self.counters.check_and_update(counters, delta)
+    }
+
+    pub fn get_counters(&self, namespace: &Namespace) -> Result<HashSet<Counter>, StorageErr> {
+        let limits = self.get_limits(namespace);
+        self.counters.get_counters(limits)
+    }
+
+    pub fn clear(&self) -> Result<(), StorageErr> {
+        self.limits.write().unwrap().clear();
+        self.counters.clear()
+    }
+}
+
+impl Default for Storage {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl AsyncStorage {
```

Comment:
Just for curiosity. Does it make sense to keep both async and sync storage version? If I am not wrong it is all about using the redis sync or async client. Shouldn't we choose one and stick to it?

Reply:
ok, got it... I think. 

Thanks for the explanation